### DISCUSSION TITLE
add "goto" like in c++ to gdscript

### DISCUSSION THREAD

**Discussion by enJey009** [godotengine/godot-proposals#6400] [Scripting]
add "goto" like in c++ to gdscript
![image](https://user-images.githubusercontent.com/99175830/222496706-0319911f-8415-469b-8691-2eadf0d06b26.png)

**Comment by AThousandShips** [comment]
This isn't a proposal, you need to fill out the different details, also a really bad idea in general

**Comment by vpellen** [comment]
Okay, one, bruh
Two, goto is disruptive enough that there's an argument for its inclusion being actively detrimental, which is not something I say often
Is there a particular issue you're having that goto would solve?

**Comment by AThousandShips** [comment]
GDScript does not have any advanced optimization or analysis *yet* but it still would make scope management a nightmare, but goto generally makes code much harder to read

**Comment by enJey009** [comment]
> Okay, one, bruh Two, goto is disruptive enough that there's an argument for its inclusion being actively detrimental, which is not something I say often Is there a particular issue you're having that goto would solve?

there was a problem, i could solve easier then i did, now i will check my code...

**Comment by enJey009** [comment]
for example
i have this code

up. sorry code didnt tab 

![Снимок экрана 2023-03-02 102446](https://user-images.githubusercontent.com/99175830/222359308-84d8daa6-5922-4649-b1ec-69cadc48bc32.png)

**Comment by AThousandShips** [comment]
C++ has the most liberal goto statement of any high level language, and the only other mainstream one I can find is C# and it doesn't allow it outside of the current scope, why should GDScript go against the flow?

So since this is a feature proposal, how should goto handle various contexts? What goto should be permitted? How should error handling look?
```gdscript
var a
first:
var b
if ...:
    var c
    second:
    var d
goto second # Should this be allowed? How should it be treated?
```

Etc. etc.

**Comment by enJey009** [comment]
> ```gdscript
> Should this be allowed? How should it be treated?
> ```

hm, i see its impossible for gdscript, because of the architecture.(but iam not sure)
GDscript devs can make something like "passall" but this "passall" will pass all IFs local code in, like in my example above.

sorry for, but usually i speak russian

**Comment by dalexeev** [comment]
In JavaScript, this is possible (break from block, not `goto`). Docs: [block](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/block), [label](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/label), [break](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/break).

```javascript
my_label:
{
    console.log("A");
    if (true) {
        break my_label;
    }
    console.log("B");
}
console.log("C");

// Prints:
// A
// C
```

In GDScript it might look like this:

```gdscript
:my_label
block: # I couldn't think of a better name.
    print("A")
    if true:
        break my_label
    print("B")
print("C")
```

Break from nested loop:

```gdscript
:outer
for i in 10:
    :inner
    for j in 10:
        if true:
            break outer
```

**Comment by AThousandShips** [comment]
The break from loops, or nested breaks like `break break` has been suggested before I think, and is a feature I would like myself

**Comment by anvilfolk** [comment]
See #3780 :)

**Comment by YuriSizov** [comment]
Please, don't make low quality posts like that. If you want to open a discussion, then open a discussion. A proposal must have more thought put into it than just replicating the same sentence 10 times.

**Reply by enJey009** [reply]
i am not professional github user. no russian lang

**Reply by AThousandShips** [reply]
What has that got to do with it? Were you unable to read the instructions?

**Comment by SlugFiller** [comment]
It's been said by others, but `goto` in C/C++ is just a poor workaround for the lack of a labelled `break`. In fact, it is used for specifically that purpose across the Godot's core code. If you're looking for a labelled `break`, you should ask for a labelled `break`. Not a `goto`. I might also add labelled `continue`, although that can be worked around using clever scoping and labelled `break`.

**Reply by Calinou** [reply]
https://github.com/godotengine/godot-proposals/issues/3780 covers labeled `break`.
